src/GPU_API/CUAPI_Asyn_FluidSolver.cu

```diff
@@ -289,47 +334,62 @@ void CUAPI_Asyn_FluidSolver( real h_Flu_Array_In[][FLU_NIN ][ CUBE(FLU_NXT) ],
 #  endif
 #  endif // #ifdef GAMER_DEBUG
 
-#  if ( !( MODEL == ELBDM && WAVE_SCHEME == WAVE_GRAMFE ) )
+#  if ( !( MODEL == ELBDM && WAVE_SCHEME == WAVE_GRAMFE && ELBDM_SCHEME != ELBDM_HYBRID && GRAMFE_SCHEME == GRAMFE_FFT) )
    const dim3 BlockDim_FluidSolver ( FLU_BLOCK_SIZE_X, FLU_BLOCK_SIZE_Y, 1 ); // for the fluidsolvers
-#  endif // #  if ( !( MODEL == ELBDM && WAVE_SCHEME == WAVE_GRAMFE ) )
+#  endif // #  if ( !( MODEL == ELBDM && WAVE_SCHEME == WAVE_GRAMFE && ELBDM_SCHEME != ELBDM_HYBRID && GRAMFE_SCHEME == GRAMFE_FFT) )
 
 // model-dependent operations
 #  if   ( MODEL == HYDRO )
 
 #  elif ( MODEL == ELBDM )
 
+#  if ( WAVE_SCHEME == WAVE_GRAMFE && GRAMFE_SCHEME == GRAMFE_FFT )
+   uint cufftdx_shared_memory_size;
+#  endif // #  #  if ( WAVE_SCHEME == WAVE_GRAMFE && GRAMFE_SCHEME == GRAMFE_FFT )
+
+#  if ( ELBDM_SCHEME == ELBDM_HYBRID )
+   if ( UseWaveFlag ) {
+#  endif // # if ( ELBDM_SCHEME == ELBDM_HYBRID )
+
 #  if ( WAVE_SCHEME == WAVE_FD )
+
 // evaluate the optimized Taylor expansion coefficient
    if ( ELBDM_Taylor3_Auto )  ELBDM_Taylor3_Coeff = ELBDM_SetTaylor3Coeff( dt, dh, ELBDM_Eta );
+
 #  elif ( WAVE_SCHEME == WAVE_GRAMFE )
-// set up GPU FFT if GPU is used for Gram Fourier extension scheme
-#  ifdef GRAMFE_ENABLE_GPU
+
+// set up GPU FFT if GPU is used for Gram Fourier extension FFT scheme
+#  if ( GRAMFE_SCHEME == GRAMFE_FFT )
 // total size of shared memory required for storing FFT::ffts_per_block rows of data after Gram extension and the coefficients of the respective left and right extension polynomials
    auto size                       = FFT::ffts_per_block * cufftdx::size_of<FFT>::value + 2 * FFT::ffts_per_block * GRAMFE_NDELTA;
    auto size_bytes                 = size * sizeof(complex_type);
 
 // shared memory must fit input data and must be big enough to run FFT
-   uint cufftdx_shared_memory_size = std::max((unsigned int)FFT::shared_memory_size, (unsigned int)size_bytes);
+   cufftdx_shared_memory_size = std::max((unsigned int)FFT::shared_memory_size, (unsigned int)size_bytes);
 
 // increase max shared memory if needed
    CUDA_CHECK_ERROR(cudaFuncSetAttribute(
-      CUFLU_ELBDMSolver_GramFE,
+      CUFLU_ELBDMSolver_GramFE_FFT,
       cudaFuncAttributeMaxDynamicSharedMemorySize,
       cufftdx_shared_memory_size));
 
-// create forward and backward cufftx workspaces
-   cudaError_t error_code  = cudaSuccess;
-   auto cufftdx_workspace  = cufftdx::make_workspace<FFT>(error_code);
-   CUDA_CHECK_ERROR(error_code);
-   error_code              = cudaSuccess;
-   auto cufftdx_iworkspace = cufftdx::make_workspace<IFFT>(error_code);
-   CUDA_CHECK_ERROR(error_code);
 
-#  endif // #  ifdef GRAMFE_ENABLE_GPU
+#  elif ( GRAMFE_SCHEME == GRAMFE_MATMUL )
+
+   size_t h_FluTimeEvo_MemSize = 2 * FLU_NXT * PS2 * sizeof(gramfe_matmul_float);
+   ELBDM_GramFE_ComputeTimeEvolutionMatrix(h_GramFE_TimeEvo, dt, dh, ELBDM_Eta);
```

Comment:
@hyschive 
One question about this part: 
I moved the call to `ELBDM_GramFE_ComputeTimeEvolutionMatrix` to `InvokeSolver` since it is indeed very expensive (and if I could use 256 bit precision for this calculation like I would like to, it would be even more expensive). 
In principle, one also needs to send this array to the GPU only once. Are you fine with moving transferring the matrix to the GPU in InvokeSolver as well? Or do you think that transferring such a small matrix (4 kilobytes for patch size 8) is negligible?